## Data Cleaning & Feature Engineering

Let's start cleaning our data. We will start this task by first looking at our data and then decide what we need to do. During this phase we will also assess if we need split or drop existing features, or create entirely new features.

For cleaning the data, I have created a .py containing all the functions that will help me clean. There can be other approaches, but this is what I'm going to follow.

A typical data cleaning task starts with reviewing the data which we have. So, let's start exploring our data and see what we have to clean.

In [1]:
#importing libraries

import pandas as pd
import data_cleaning as dc

In [2]:
# reading our properties data into dataframe

df_properties = pd.read_csv('properties.csv')

# lets see the dimensions of the data
print('We have', df_properties.shape[0], 'rows and', df_properties.shape[1], 'columns in our dataframe')


We have 1989 rows and 12 columns in our dataframe


In [3]:
df_properties.head(3)

id                                              title    price  \
0  5528049  Price Reduced | Middle Floor | Nice Sea View | PJ  2700000   
1  6008529       2BR Type D | Sea and Burj Al Arab Views | PJ  2850000   
2  6034542  AMAZING LAKE VIEW / 3 BR w/ Maid's Rm Apt @ JU...  1150000   

                                       location   latitude  longitude  \
0                          Dubai, Palm Jumeirah  25.113208  55.138932   
1    Dubai, Palm Jumeirah, Shoreline Apartments  25.106809  55.151201   
2  Dubai, Jumeirah Lake Towers, Lake Almas West  25.063302  55.137728   

        type                                               size  \
0  Apartment  1,079 sqft\n                      /           ...   
1  Apartment  1,582 sqft\n                      /           ...   
2  Apartment  1,951 sqft\n                      /           ...   

  no_of_bedrooms  no_of_bathrooms completion_status  \
0              1              2.0          Off-plan   
1              2              2.0             Ready   
2       3 + Maid              5.0               NaN   

                                           amenities  
0  Partly furnished, Balcony, Barbecue Area, Cent...  
1  Furnished, Balcony, Built in Wardrobes, Centra...  
2  Unfurnished, Balcony, Built in Wardrobes, Cove...

#### 1. Removing unwanted columns

Looking at the above data, we can see that we have a title and type columns, these does not add much value to our analysis and modeling. So we can safely remove them from our dataframe.

All other columns are relevant to either model building or exploratory analysis.

In [4]:
# calling our function to remove unwanted columns
column_list = ['title', 'type']
df_properties = dc.remove_unwanted_columns(df_properties, column_list)

# lets preview our data again after deletion
df_properties.head()

id    price                                      location   latitude  \
0  5528049  2700000                          Dubai, Palm Jumeirah  25.113208   
1  6008529  2850000    Dubai, Palm Jumeirah, Shoreline Apartments  25.106809   
2  6034542  1150000  Dubai, Jumeirah Lake Towers, Lake Almas West  25.063302   
3  6326063  2850000                        Dubai, Culture Village  25.227295   
4  6356778  1729200                          Dubai, Palm Jumeirah  25.114275   

   longitude                                               size  \
0  55.138932  1,079 sqft\n                      /           ...   
1  55.151201  1,582 sqft\n                      /           ...   
2  55.137728  1,951 sqft\n                      /           ...   
3  55.341761  2,020 sqft\n                      /           ...   
4  55.139764  507 sqft\n                      /            4...   

  no_of_bedrooms  no_of_bathrooms completion_status  \
0              1              2.0          Off-plan   
1              2              2.0             Ready   
2       3 + Maid              5.0               NaN   
3              2              3.0             Ready   
4         studio              1.0          Off-plan   

                                           amenities  
0  Partly furnished, Balcony, Barbecue Area, Cent...  
1  Furnished, Balcony, Built in Wardrobes, Centra...  
2  Unfurnished, Balcony, Built in Wardrobes, Cove...  
3  Unfurnished, Balcony, Concierge, Covered Parki...  
4  Furnished, Built in Wardrobes, Central A/C, Co...

#### 2. Removing duplicates

Let's make sure we don't have any duplicate properties. To do that, let's look at our id field for duplication.

In [5]:
pd.set_option('display.max_rows', 25)

property_ids = df_properties['id']
df_properties[property_ids.isin(property_ids[property_ids.duplicated()])].sort_values('id')

id    price                                           location  \
8     6376886  2100000         Dubai, Palm Jumeirah, Shoreline Apartments   
9     6376886  2100000         Dubai, Palm Jumeirah, Shoreline Apartments   
11    6473651  3550000                  Dubai, Downtown Dubai, Burj Vista   
12    6473651  3550000                  Dubai, Downtown Dubai, Burj Vista   
28    6810606  2800000           Dubai, Downtown Dubai, Burj Khalifa Area   
...       ...      ...                                                ...   
1846  7695774  1366000                            Dubai, Jumeirah, La Mer   
1895  7698639  2400000                    Dubai, DIFC, Central Park Tower   
1896  7698639  2400000                    Dubai, DIFC, Central Park Tower   
1975  7704274  1175000  Dubai, Dubai South (Dubai World Central), EMAA...   
1976  7704274  1175000  Dubai, Dubai South (Dubai World Central), EMAA...   

       latitude  longitude                                               size  \
8     25.106668  55.149275  2,186 sqft\n                      /           ...   
9     25.106668  55.149275  2,186 sqft\n                      /           ...   
11    25.198796  55.271342  1,918 sqft\n                      /           ...   
12    25.198796  55.271342  1,918 sqft\n                      /           ...   
28    25.194282  55.271884  1,712 sqft\n                      /           ...   
...         ...        ...                                                ...   
1846  25.240419  55.252770  816 sqft\n                      /            7...   
1895  25.206412  55.275199  1,782 sqft\n                      /           ...   
1896  25.206412  55.275199  1,782 sqft\n                      /           ...   
1975  24.865992  55.137958  3,247 sqft\n                      /           ...   
1976  24.865992  55.137958  3,247 sqft\n                      /           ...   

     no_of_bedrooms  no_of_bathrooms completion_status  \
8                 3              3.0             Ready   
9                 3              3.0             Ready   
11                3              4.0             Ready   
12                3              4.0             Ready   
28                2              3.0          Off-plan   
...             ...              ...               ...   
1846              1              1.0          Off-plan   
1895              2              3.0             Ready   
1896              2              3.0             Ready   
1975              3              2.0             Ready   
1976              3              2.0             Ready   

                                              amenities  
8                                                   NaN  
9                                                   NaN  
11    Unfurnished, Balcony, Central A/C, Kitchen App...  
12    Unfurnished, Balcony, Central A/C, Kitchen App...  
28    Unfurnished, Balcony, Built in Wardrobes, Cent...  
...                                                 ...  
1846  Unfurnished, Balcony, Built in Wardrobes, Cent...  
1895                                                NaN  
1896                                                NaN  
1975  Partly furnished, Balcony, Barbecue Area, Buil...  
1976  Partly furnished, Balcony, Barbecue Area, Buil...  

[112 rows x 10 columns]

We can see above that we have 112 duplicate properties. Which means we have to remove 56 rows out of our properties data

In [6]:
# calling our custom function to remove duplicate properties on the basis of id

df_properties = dc.remove_duplocate_properties(df_properties)
property_ids = df_properties['id']
df_properties[property_ids.isin(property_ids[property_ids.duplicated()])].sort_values('id')

Empty DataFrame
Columns: [id, price, location, latitude, longitude, size, no_of_bedrooms, no_of_bathrooms, completion_status, amenities]
Index: []

In [7]:
# As we can see now, we don't have any duplicates. Let's see the dimensions of our data

print('We have', df_properties.shape[0], 'rows and', df_properties.shape[1], 'columns in our dataframe')

We have 1932 rows and 10 columns in our dataframe


Comparing to our initial number of properties 1992, we now have 1935 unique properties. Which means we removed 57 redundent rows from our dataframe

#### 4. Removing NULL and unwanted values

Now let's see where we need to remove NULL data or unwanted data from our dataframe. 

To do this, we will first try to find our what data we have in each columns and then make a judgement on how to proceed with this task

##### 4.1 Price

In [8]:
# Unique values for price column

df_properties['price'].value_counts()

1300000          31
1650000          29
1200000          28
Ask for price    27
1100000          26
                 ..
954505            1
1502000           1
1276686           1
835000            1
4222888           1
Name: price, Length: 822, dtype: int64

In [9]:
# As we can see above that some of the properties have their price defined as 'Ask for price'. This is certinly not going to help us. So let's drop all the properties where price is defined as 'Ask for price'

df_properties = dc.remove_askforprice_properties(df_properties)
df_properties['price'].value_counts()

1300000    31
1650000    29
1200000    28
1100000    26
2100000    24
           ..
1276686     1
835000      1
2299900     1
1439112     1
4222888     1
Name: price, Length: 821, dtype: int64

As you can see above, we have removed all the properties having price as 'Ask for price'

##### 4.2 Number of Bathrooms

In [10]:
df_properties['no_of_bathrooms'].value_counts()

2.0    704
3.0    553
1.0    311
4.0    230
5.0     77
6.0     11
Name: no_of_bathrooms, dtype: int64

In [11]:
# Now, before handline 19 NULL values in no_of_bathrooms, we will need to perform certain steps.

# fill nan with 0 in no_of_bathrooms

replacement= {'no_of_bathrooms': 0} 
df_properties = dc.replace_nan(df_properties, replacement)

# As no_of_bathrooms column is an object, lets convert it to int to get mean and average values
convert_dict = {'no_of_bathrooms' : int}

df_properties = dc.convert_datatypes(df_properties, convert_dict)

In [12]:
# now that bathrooms column is converted to int and null values are replaced with. Let's find out what is the mean value for bathroom and then use that value to replace NULL

print("Mean:", df_properties['no_of_bathrooms'].mean())
print("Rounded Mean number of bathrooms:", df_properties['no_of_bathrooms'].mean().round())

Mean: 2.4929133858267716
Rounded Mean number of bathrooms: 2.0


In [13]:
# Now lets replace 0 value with our mean value of 2

df_properties['no_of_bathrooms'].replace(0, 2, inplace = True)

df_properties['no_of_bathrooms'].value_counts()

2    723
3    553
1    311
4    230
5     77
6     11
Name: no_of_bathrooms, dtype: int64

##### 4.3 Number of Bedrooms

In [14]:
df_properties['no_of_bedrooms'].value_counts()

2                646
1                596
3                220
studio           122
3 + Maid         118
2 + Maid          83
1 + Maid          43
4                 42
4 + Maid          28
5                  4
studio + Maid      2
5 + Maid           1
Name: no_of_bedrooms, dtype: int64

As we can see above, bedroom column includes variety of data. So:
- '+ Maid' will be removed from bedroom column and converted into a new feature call MaidRoom 0 or 1, indicating whether the apartment has a maid room or not
- studio will be converted to 0, as their really isn't any separate bedrooms in studio apartment

In [15]:
df_properties = dc.fix_bedroom_column(df_properties)
df_properties['id'].value_counts()

7579647    1
7523715    1
7652712    1
7652714    1
7617899    1
          ..
7690950    1
7602887    1
7498440    1
7588554    1
7499776    1
Name: id, Length: 1905, dtype: int64

##### 4.4 Amenities

For handling amenities, we will do:
1. Retreive the amenities value from the apartment which has most number of amenities defined, by doing MAX(string). That will represent the maximum number of amenities an apartment can have.
2. Clean up the retreived amenities list of special chars, spaces, etc
3. Split the amenities list into individual feature with default value of 0
4. Loop all properties and set each amenity feature to either 1 or 0 if that apartment has that perticular amenity. I hope this makes sense :)

In [16]:
df_properties['amenities'] = df_properties['amenities'].astype(str)
most_amenities = max(df_properties['amenities'], key = len)
most_amenities

"Unfurnished, Balcony, Barbecue Area, Built in Wardrobes, Central A/C, Children's Play Area, Children's Pool, Concierge, Covered Parking, Kitchen Appliances, Lobby in Building, Maid Service, Networked, Pets Allowed, Private Garden, Private Gym, Private Jacuzzi, Private Pool, Security, Shared Gym, Shared Pool, Shared Spa, Study, Vastu-compliant, View of Landmark, View of Water, Walk-in Closet, "

In [17]:
# Retreiving MAX amenity from dataframe
df_properties['amenities'] = df_properties['amenities'].astype(str)
most_amenities = max(df_properties['amenities'], key = len)

# Custom function for converting amenities to column
df_properties, new_amenities_list = dc.convert_amenities_to_columns(df_properties, most_amenities)

# Setting 1 or 0 against each property for their amenity
df_properties = dc.fill_amenities(df_properties, new_amenities_list)

df_properties.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1905 entries, 0 to 1988
Data columns (total 38 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   id                   1905 non-null   int64  
 1   price                1905 non-null   object 
 2   location             1905 non-null   object 
 3   latitude             1905 non-null   float64
 4   longitude            1905 non-null   float64
 5   size                 1905 non-null   object 
 6   no_of_bedrooms       1905 non-null   object 
 7   no_of_bathrooms      1905 non-null   int32  
 8   completion_status    1740 non-null   object 
 9   amenities            1905 non-null   object 
 10  maid_room            1905 non-null   bool   
 11  unfurnished          1905 non-null   bool   
 12  balcony              1905 non-null   bool   
 13  barbecue_area        1905 non-null   bool   
 14  built_in_wardrobes   1905 non-null   bool   
 15  central_ac           1905 non-null   b

##### 4.5 Apartment Quality

It is safe to say that based on number of amenities offered in an apartment, we can set the quality for each apartment.

But do note that we might miscatgorize the apartment just because the amenities are not recorded by property portal, even though it's highly priced apartment.

But as we are here for learning, so lets do this.

In [18]:
print('Maxiumum number of amenities any apartment has is ', len(most_amenities.split(', ')))

Maxiumum number of amenities any apartment has is  28


Based on above number we can rate our properties into following categories:
    
(1 - 7): Low<br>
(8 - 14): Medium<br>
(15 - 21): High<br>
(22 - 28): Ultra

In [19]:
df_properties = dc.mark_property_quality(df_properties)
df_properties.quality.value_counts()

Medium    1146
Low        544
High       134
Ultra       81
Name: quality, dtype: int64

As you can see we have successfully engineered a categorical feature for our property dataset

##### 4.6 Size

Let's lookinto size feature of our dataframe and clean-up the data. We will also engineer a new feature called price_per_sqft. This will be usefull for our exploratory analysis to see which localities are more expensive than others in terms of property pricings

In [20]:
df_properties['size'].value_counts()

1,400 sqft\n                      /            130 sqm    10
1,300 sqft\n                      /            121 sqm    10
851 sqft\n                      /            79 sqm        9
1,133 sqft\n                      /            105 sqm     8
895 sqft\n                      /            83 sqm        7
                                                          ..
5,049 sqft\n                      /            469 sqm     1
720 sqft\n                      /            67 sqm        1
1,406 sqft\n                      /            131 sqm     1
294 sqft\n                      /            27 sqm        1
1,360 sqft\n                      /            126 sqm     1
Name: size, Length: 1121, dtype: int64

As you can see, size column contains details of both SQFT and SQM. For our analysis, we will keep the SQFT details and discard the SQM data, As keeping both values dosn't add value to our analysis

In [21]:
# Cleaning up property size details using custom function
df_properties = dc.cleanup_property_size(df_properties)

# rename column
df_properties.rename(columns={'size':'size_in_sqft'}, inplace = True)

# lets create a new computed
df_properties = dc.create_feature_price_per_sqft(df_properties)

df_properties.head(3)

id                                      location   latitude  \
0  5528049                          Dubai, Palm Jumeirah  25.113208   
1  6008529    Dubai, Palm Jumeirah, Shoreline Apartments  25.106809   
2  6034542  Dubai, Jumeirah Lake Towers, Lake Almas West  25.063302   

   longitude    price  size_in_sqft  price_per_sqft no_of_bedrooms  \
0  55.138932  2700000          1079         2502.32              1   
1  55.151201  2850000          1582         1801.52              2   
2  55.137728  1150000          1951          589.44              3   

   no_of_bathrooms quality  ... private_pool security  shared_gym  \
0                2  Medium  ...        False    False        True   
1                2  Medium  ...        False    False        True   
2                5  Medium  ...        False     True        True   

   shared_pool  shared_spa  study  vastu_compliant  view_of_landmark  \
0        False       False  False            False             False   
1         True       False  False            False             False   
2         True       False  False            False              True   

   view_of_water  walk_in_closet  
0           True           False  
1           True           False  
2           True            True  

[3 rows x 40 columns]

##### 4.7 Completion Status

Let's take a look at completion status field and see what we find there

In [22]:
# Lets see what unique values we have.

df_properties.completion_status.value_counts()

Ready       1132
Off-plan     608
Name: completion_status, dtype: int64

In [23]:
# Lets count the null values

df_properties.completion_status.isna().sum()

165

As we can see, we have significant amount of properties without completion status. This can taint the findings.

One way of handling this is to replace the empty completion status with 'Ready', as most of the properties are marked as ready.

Other way, is to eliminate the properties which have no completion status.

If we assess the completion status field, we can say that it is not that significant of a field to have the properties removed for it. So best way forward is to drop this column.

In [24]:
# Dropping completion_status field

df_properties.drop('completion_status', axis = 1, inplace = True)

In real world, the decision of dropping a column should be based on stakeholder input. We are just practicing, so why not :)

##### 4.8 Location

In [25]:
df_properties.location.value_counts()

Dubai, Jumeirah Village Circle                           183
Dubai, Dubai Marina                                      160
Dubai, Downtown Dubai                                     90
Dubai, Business Bay                                       60
Dubai, Palm Jumeirah                                      56
                                                        ... 
Dubai, Dubai Marina, Marina Residence                      1
Dubai, Dubai South (Dubai World Central), EMAAR South      1
Dubai, Dubai Silicon Oasis, La Vista Residence             1
Dubai, Jumeirah Village Circle, District 18                1
Dubai, Dubai Sports City, Zenith Towers                    1
Name: location, Length: 217, dtype: int64

If we look above, we can see that in some cases we have only have location as "Dubai, [Neighborhood Name]" and in same case we have "Dubai, [Neighborhood Name], Building/Project Name"

Let's standardize this by:

- Remove "Dubai, " part from location as we all know all properties are from Dubai
- Remove the project part of Location as this analysis is more targeted to locality rather than project or building
- All locations containing word downtown in neighbor, except for the ones called downtown jebel ali

In [26]:
df_properties = dc.clean_location_details(df_properties)

In [27]:
df_properties.neighborhood.unique()

array(['Palm Jumeirah', 'Jumeirah Lake Towers', 'Culture Village',
       'Downtown Dubai', 'Dubai Marina', 'Business Bay', 'Old Town',
       'Al Kifaf', 'Meydan', 'Arjan', 'Jumeirah Beach Residence',
       'Dubai Creek Harbour (The Lagoons)', 'Greens', 'City Walk',
       'Al Furjan', 'DAMAC Hills', 'Jumeirah Golf Estates', 'Jumeirah',
       'Dubai Hills Estate', 'Umm Suqeim', 'Motor City', 'DIFC',
       'Jumeirah Village Circle', 'Barsha Heights (Tecom)', 'Al Barari',
       'Dubai Production City (IMPZ)', 'The Hills', 'The Views',
       'Dubai Sports City', 'Dubai Silicon Oasis',
       'Jumeirah Village Triangle', 'Mohammed Bin Rashid City',
       'Dubai Harbour', 'Bluewaters', 'International City',
       'Falcon City of Wonders', 'Mina Rashid', 'Town Square',
       'Green Community', 'Al Barsha', 'Al Sufouh', 'Dubai Festival City',
       'Jebel Ali', 'Dubai Land', 'World Trade Center', 'Mudon',
       'Discovery Gardens', 'Remraam', 'Mirdif',
       'Dubai South (Dubai Wo

In [28]:
df_properties.neighborhood.value_counts()

Downtown Dubai              302
Dubai Marina                288
Jumeirah Village Circle     200
Palm Jumeirah               178
Jumeirah Beach Residence    116
                           ... 
Dubai Healthcare City         2
Dubai Residence Complex       1
Al Quoz                       1
Mina Rashid                   1
wasl gate                     1
Name: neighborhood, Length: 54, dtype: int64

Now we have a very normalized neighborhood data which can give us helpful insight of price trends per Dubai neighborhoods

#### 5. Saving Data

Now that we have cleaned our data, it's time to save this updated dataset into CSV format.

We will utilize this updated CSV in our exploratory and machine learning analysis

In [29]:
df_properties.to_csv('cleaned_property_data.csv', index = False)